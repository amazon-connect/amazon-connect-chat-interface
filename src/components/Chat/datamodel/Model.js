// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

export const PARTICIPANT_MESSAGE = "MESSAGE";
export const CHAT_EVENT = "EVENT";
export const ATTACHMENT_MESSAGE = "ATTACHMENT";

export const PARTICIPANT_TYPES = {
    AGENT: "AGENT",
    CUSTOMER: "CUSTOMER"
};

export const ContentType = {
    EVENT_CONTENT_TYPE: {
        TYPING: "application/vnd.amazonaws.connect.event.typing",
        PARTICIPANT_JOINED: "application/vnd.amazonaws.connect.event.participant.joined",
        PARTICIPANT_LEFT: "application/vnd.amazonaws.connect.event.participant.left",
        TRANSFER_SUCCEEDED: "application/vnd.amazonaws.connect.event.transfer.succeeded",
        TRANSFER_FAILED: "application/vnd.amazonaws.connect.event.transfer.failed",
        CONNECTION_ACKNOWLEDGED: "application/vnd.amazonaws.connect.event.connection.acknowledged",
        CHAT_ENDED: "application/vnd.amazonaws.connect.event.chat.ended"
    },
    MESSAGE_CONTENT_TYPE: {
        TEXT_PLAIN: "text/plain",
        APPLICATION_PDF: "application/pdf",
        IMAGE_JPG: "image/jpeg",
        IMAGE_PNG: "image/png",
        APPLICATION_DOC: "application/msword",
        APPLICATION_DOCX: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        APPLICATION_XLS: "application/vnd.ms-excel",
        APPLICATION_XLSX: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        TEXT_CSV: "text/csv",
        APPLICATION_PPT: "application/vnd.ms-powerpoint",
        APPLICATION_PPTX: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        AUDIO_WAV: "audio/wav",
        AUDIO_X_WAV: "audio/x-wav", //Firefox
        AUDIO_VND_WAVE: "audio/vnd.wave", //IE
        INTERACTIVE_MESSAGE: "application/vnd.amazonaws.connect.message.interactive"
    }
};

ContentType.ATTACHMENT_CONTENT_TYPE = {
    TXT: ContentType.MESSAGE_CONTENT_TYPE.TEXT_PLAIN,
    PDF: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PDF,
    JPG: ContentType.MESSAGE_CONTENT_TYPE.IMAGE_JPG,
    PNG: ContentType.MESSAGE_CONTENT_TYPE.IMAGE_PNG,
    DOC: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_DOC,
    DOCX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_DOCX,
    XLS: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_XLS,
    XLSX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_XLSX,
    CSV: ContentType.MESSAGE_CONTENT_TYPE.TEXT_CSV,
    PPT: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PPT,
    PPTX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PPTX,
    WAV: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_WAV,
    X_WAV: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_X_WAV,
    VND_WAVE: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_VND_WAVE
};

//OpenXML content types do not show up in custom files list unless file extension is explicitly provided
export const ATTACHMENT_ACCEPT_CONTENT_TYPES = [
    ...Object.values(ContentType.ATTACHMENT_CONTENT_TYPE),
    ".docx",
    ".xlsx",
    ".pptx"
];

export const InteractiveMessageType = {
    LIST_PICKER: "ListPicker",
    TIME_PICKER: "TimePicker",
    PANEL: "Panel"
};

export const Status = {
    Sending: "Sending",
    SendSuccess: "SendSuccess",
    SendFailed: "SendFailed",
    Read: "Read"
    // Delivered?
};

export const AttachmentStatus = {
    APPROVED: "APPROVED",
    REJECTED: "REJECTED"
};

export const Direction = {
    Outgoing: "Outgoing",
    Incoming: "Incoming"
};

export const AttachmentErrorType = {
    AccessDeniedException: "AccessDeniedException",
    InternalServerException: "InternalServerException",
    ThrottlingException: "ThrottlingException",
    ValidationException: "ValidationException",
    ServiceQuotaExceededException: "ServiceQuotaExceededException",
    ResourceNotFoundException: "ResourceNotFoundException",
    ConflictException: "ConflictException"
};
export class TransportDetails {
    constructor(input) {
        var args = input || {};
        this.direction = args.direction;
        this.readTime = args.readTime;
        this.status = args.status;
        this.sentTime = args.sentTime;
    }
}

export class ItemDetails {
    constructor(input) {
        var args = input || {};
        this.id = args.id;
        this.type = args.type;
        this.content = args.content;
        this.displayName = args.displayName;
        this.participantId = args.participantId;
        this.participantRole = args.participantRole;
        this.transportDetails = args.transportDetails;
    }
}
