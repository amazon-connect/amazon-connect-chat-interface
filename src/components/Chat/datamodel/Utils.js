// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import {
  Status,
  Direction,
  ItemDetails,
  TransportDetails,
  ContentType,
  PARTICIPANT_TYPES
} from "./Model";

function isRecognizedEvent(eventName) {
  var values = Object.values(ContentType.EVENT_CONTENT_TYPE);
  for (var index = 0; index < values.length; index++) {
    if (values[index] === eventName) {
      return true;
    }
  }
  return false;
}

function getContent(item) {
  var content = {};
  content.data = item.Content;
  content.type = item.ContentType;
  return content;
}

function createItemFromIncoming(item, thisParticipant) {
  var transcriptItem = new ItemDetails();
  var transportDetails = new TransportDetails();
  transcriptItem.id = item.Id;
  transcriptItem.type = item.Type;
  transcriptItem.content = getContent(item);
  transcriptItem.displayName = item.DisplayName;
  transcriptItem.participantId = item.ParticipantId;
  transcriptItem.participantRole = item.ParticipantRole;
  transportDetails.direction = 
      item.ParticipantRole === PARTICIPANT_TYPES.CUSTOMER
        ? Direction.Outgoing
        : Direction.Incoming;
  transportDetails.sentTime = 
    new Date(item.AbsoluteTime).getTime() / 1000;
  transportDetails.status = Status.SendSuccess;
  transcriptItem.transportDetails = transportDetails;
  transcriptItem.version = 0;
  transcriptItem.Attachments = item.Attachments;
  return transcriptItem;
}

function createTranscriptItemFromSuccessResponse(oldTranscriptItem, response) {
  const newTranscriptItem = new ItemDetails();
  Object.assign(newTranscriptItem, oldTranscriptItem);

  if(response.data && response.data.Id){
    newTranscriptItem.id = response.data.Id;
  }

  newTranscriptItem.transportDetails = {
    ...oldTranscriptItem.transportDetails,
    status: Status.SendSuccess,
    sentTime: new Date(response.data.AbsoluteTime || Date.now()).getTime() / 1000
  };
  return newTranscriptItem;
}

function createOutgoingTranscriptItem(type, content, participant){
  const transcriptItem = new ItemDetails();
  const transportDetails = {};
  transcriptItem.type = type;
  transcriptItem.content = content;
  transcriptItem.participantId = participant.participantId;
  transcriptItem.participantRole = PARTICIPANT_TYPES.CUSTOMER;
  transcriptItem.displayName = participant.displayName;
  transportDetails.status = Status.Sending;
  transportDetails.direction = Direction.Outgoing;
  transportDetails.sentTime = _timestampNow();
  transcriptItem.transportDetails = transportDetails;
  transcriptItem.id = _generateLocalId();
  transcriptItem.version = 0;
  return transcriptItem;
}

function createFailedItem(item, sentTime) {
  const clonedItem = new ItemDetails(item);
  clonedItem.transportDetails.status = Status.SendFailed;
  clonedItem.transportDetails.sentTime = sentTime;
  clonedItem.version = clonedItem.version + 1;
  return clonedItem;
}

function _generateLocalId() {
  var dt = new Date().getTime();
  var uuid = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(
    c
  ) {
    var r = (dt + Math.random() * 16) % 16 | 0;
    dt = Math.floor(dt / 16);
    return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
  });
  return uuid;
}

function _timestampNow() {
  var date = new Date();
  var timestamp = date.getTime();
  return timestamp / 1000;
}

function createTypingParticipant(typingDataItem, thisParticipantId) {
  var participantTypingDetails = {};
  participantTypingDetails.participantId = typingDataItem.ParticipantId;
  participantTypingDetails.displayName = typingDataItem.DisplayName;
  var direction =
    thisParticipantId === typingDataItem.ParticipantId
      ? Direction.Outgoing
      : Direction.Incoming;
  participantTypingDetails.direction = direction;
  return participantTypingDetails;
}

function isAttachmentContentType(contentType) {
  return contentType && Object.values(ContentType.ATTACHMENT_CONTENT_TYPE).includes(contentType.toLowerCase());
}

var modelUtils = {
  createItemFromIncoming: createItemFromIncoming,
  createOutgoingTranscriptItem: createOutgoingTranscriptItem,
  createFailedItem: createFailedItem,
  createTypingParticipant: createTypingParticipant,
  isRecognizedEvent: isRecognizedEvent,
  createTranscriptItemFromSuccessResponse: createTranscriptItemFromSuccessResponse,
  isAttachmentContentType: isAttachmentContentType,
};

export { modelUtils };