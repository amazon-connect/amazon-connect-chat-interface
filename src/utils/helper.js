import { ContentType, InteractiveMessageType } from "../components/Chat/datamodel/Model";
import { INTERACTIVE_MESSAGE } from "../components/Chat/constants";
import * as DOMPurify from 'dompurify';

export function shouldDisplayMessageForType(correntType) {
  let isValid = false;
  for (let key in ContentType.MESSAGE_CONTENT_TYPE) {
    if (ContentType.MESSAGE_CONTENT_TYPE[key] === correntType) {
      isValid = true;
    }
  }
  return isValid;
}

export function getTimeFromTimeStamp(timeStamp) {
  return new Date(timeStamp).getTime();
}

export function createInteractiveMessagePayload(
  selectedElement,
  preIndex,
  nextIndex,
  listId,
  templateType,
  referenceId,
  isCarouselElem = false,
  listTitle,
  carouselTemplateId
) {
  if (isCarouselElem) {
    return {
        text: JSON.stringify({
          templateIdentifier: carouselTemplateId,
          listTitle,
          selectionText: selectedElement.title
        })
    }
  }

  const payload = { text: selectedElement.title };
  if (
    selectedElement.actionDetail &&
    (selectedElement.actionDetail === INTERACTIVE_MESSAGE.ACTIONS.SHOW_MORE ||
      selectedElement.actionDetail ===
        INTERACTIVE_MESSAGE.ACTIONS.PREVIOUS_OPTIONS)
  ) {
    const requestBody = {
      version: INTERACTIVE_MESSAGE.VERSION,

      data: {
        actionName: selectedElement.actionDetail,
        preIndex,
        nextIndex,
        listId,
        templateType,
        referenceId,
      },
      action: selectedElement.actionDetail,
    };
    const jsonStr = JSON.stringify(requestBody);
    payload.text = jsonStr;
    payload.type = ContentType.MESSAGE_CONTENT_TYPE.INTERACTIVE_RESPONSE;
  }
  return payload;
}

/**
 * Frontend validations for interactive messages
 *  - Client will always receive a valid template to render
 *  - Frontend will truncate fields
 *  - Upper limits not enforced by backend, beyond a 20k char limit for the entire message
 *
 * Documentation: https://docs.aws.amazon.com/connect/latest/adminguide/interactive-messages.html
 */
export const INTERACTIVE_MESSAGE_CONSTRAINTS = {
  [InteractiveMessageType.PANEL]: {
    titleCharLimit: 400,
    subtitleCharLimit: 400,
    elementsRenderedMax: 10,
    elementTitleCharLimit: 400
  },
  [InteractiveMessageType.LIST_PICKER]: {
    titleCharLimit: 400,
    subtitleCharLimit: 400,
    elementsRenderedMax: 6,
    elementTitleCharLimit: 400,
    elementSubtitleCharLimit: 400,
  },
  [InteractiveMessageType.TIME_PICKER]: {
    titleCharLimit: 400,
    subtitleCharLimit: 400,
  },
  [InteractiveMessageType.CAROUSEL]: {
    titleCharLimit: 400,
    elementsRenderedMax: 5,
  },
  [InteractiveMessageType.QUICK_REPLY]: {
    titleCharLimit: 400,
    elementsRenderedMax: 5,
    replyOptionCharLimit: 200,
  }
}

/**
 * Truncates a string for interactive message based on hard-coded constraints
 *
 * @param {string} str - input string to truncate (e.g. title, subtitle).
 * @param {string} InteractiveMessageType - interactive message templateType.
 * @param {string} fieldLimitKey - field key for contraint (e.g. titleCharLimit).
 * @returns {string} the truncated string.
 */
export const truncateStrFromCharLimit = (str, InteractiveMessageType, fieldLimitKey) => {
  const templateContraints = INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType] || {};
  const MAX_LENGTH = templateContraints[fieldLimitKey] || 0;

  if (!(str && typeof str === "string" && MAX_LENGTH)) {
    return "";
  }

  /**
   * Mitigate Interactive Message fields XSS vulnerabilites with `dompurify`
   *
   * React will always render these passed template field values as strings not HTML.
   * React auto-escapes by default unless using dangerouslySetInnerHTML.
   */
  const sanitizedStr = DOMPurify.sanitize(str);

  if (sanitizedStr.length <= MAX_LENGTH) {
    return sanitizedStr;
  } else {
    console.warn(`[${InteractiveMessageType} template] ${fieldLimitKey} of ${MAX_LENGTH} was exceeded`)
    return sanitizedStr.substring(0, MAX_LENGTH) + "...";
  }
}

/**
 * Truncates an array of elements based on the specified minimum and maximum limits.
 *
 * @param {Array} elements - The array of elements to truncate.
 * @param {string} InteractiveMessageType - interactive message templateType.
 * @param {string} elementsRenderedMaxKey - key to retrieve max constraint from INTERACTIVE_MESSAGE_CONSTRAINTS.
 * @returns {Array} The truncated array of elements.
 */
export const truncateElementFromLimit = (elements, InteractiveMessageType, elementsRenderedMaxKey) => {
  const templateContraints = INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType] || {};
  const MAX_LIMIT = templateContraints[elementsRenderedMaxKey] || elements.length ;

  if (elements.length > MAX_LIMIT) {
    console.warn(`[${InteractiveMessageType} template] ${elementsRenderedMaxKey} of ${MAX_LIMIT} was exceeded`)
    return elements.slice(0, MAX_LIMIT);
  }

  return elements;
}
