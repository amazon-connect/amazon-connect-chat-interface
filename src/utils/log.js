/* Add your own logger function here
  var customizedLogger = {
    debug: (data) => {// customize logger function here},
    info: (data) => {// customize logger function here},
    error: (data) => {// customize logger function here}
  }
  "connect" is a global object that exported from src/components/Chat/amazon-connect-chat.js
  There are four levels available from connect.LogLevel - DEBUG, INFO, ERROR.
*/
export const config = {
    loggerConfig: {
        // Add your logger here
        // customizedLogger: customizedLogger,
        level: connect.LogLevel.INFO,
        useDefaultLogger: true
    },
};
