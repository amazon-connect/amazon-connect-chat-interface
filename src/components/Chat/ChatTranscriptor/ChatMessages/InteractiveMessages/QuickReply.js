// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from "react";
import styled from "styled-components";
import PT from "prop-types";
import { RichMessageRenderer } from "../../../RichMessageComponents";
import { Button } from "connect-core";
import { MessageBody } from "../InteractiveMessage";
import { truncateStrFromCharLimit } from "../../../../../utils/helper";
import { InteractiveMessageType } from "../../../datamodel/Model";

const customerQuickReplyStyles = {
  color: "#077398",
  background: '#FFFFFF'
}

//#region Styled Components
const ResponsesSection = styled.div`
  padding: ${({ theme }) => theme.spacing.base} 0;
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  gap: ${({ theme }) => theme.spacing.mini};
  justify-content: flex-end;
`;

const QuickReplyOption = styled(Button)`
  padding: 4px ${({ theme }) => theme.spacing.base};
  text-align: left;
  cursor: pointer;
  border: 2px solid ${customerQuickReplyStyles.color};
  border-radius: 2px;
  white-space: pre-wrap;
  font-weight: bold;
  font-size: ${({ theme }) => theme.fontsSize.regular};
  line-height: 22px;
  font-weight: 600;
  font-style: normal;
  color: ${customerQuickReplyStyles.color};
  background: ${customerQuickReplyStyles.background};
  text-shadow: none;
  box-shadow: none;

  :hover, :focus {
    background: ${customerQuickReplyStyles.color};
    color: ${customerQuickReplyStyles.background};
    border: 2px solid ${customerQuickReplyStyles.color} !important;
  }
`;
//#endregion Styled Components

function ReplyElement({ element, handleSelection }) {
  const title = truncateStrFromCharLimit( element.title, InteractiveMessageType.QUICK_REPLY, "replyOptionCharLimit");

  return (
    <QuickReplyOption onClick={() => handleSelection({ text: element.title })}>
      {title}
    </QuickReplyOption>
  );
}

QuickReply.propTypes = {
  content: PT.object.isRequired,
  addMessage: PT.func.isRequired,
};

export default function QuickReply({ content, addMessage }) {
  // assumptions: version 1. Guaranteed title exists, at least 2 elements.
  const { title: inputTitle, elements } = content;
  const title = truncateStrFromCharLimit(inputTitle, InteractiveMessageType.QUICK_REPLY, "titleCharLimit");

  return (
    <>
      <MessageBody addChildBackgroundStyles={true} data-testid="interactive-quickreply-message-title" applySpeechBubbleCaret={true}>
        <RichMessageRenderer content={title} />
      </MessageBody>
      <ResponsesSection data-testid="interactive-quickreply-response-section">
        {elements.slice(0, 5).map((element, index) => (
          <ReplyElement
            element={element}
            handleSelection={addMessage}
            key={index}
          />
        ))}
      </ResponsesSection>
    </>
  );
}
