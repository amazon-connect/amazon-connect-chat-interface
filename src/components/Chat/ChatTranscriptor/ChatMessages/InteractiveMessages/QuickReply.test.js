// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from "react";
import { screen, render, fireEvent } from "@testing-library/react";
import { ThemeProvider } from "../../../../../theme";
import QuickReply from "./QuickReply";
import { InteractiveMessageType } from "../../../datamodel/Model";
import * as helpers from '../../../../../utils/helper';

const QUICK_REPLY_CONSTRAINTS = helpers.INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType.QUICK_REPLY];

export const mockQuickReplyContent = {
  title: "How was your experience?",
  elements: [
    {
      title: "Great!",
    },
    {
      title: "Good",
    },
    {
      title: "Ok",
    },
    {
      title: "Poor",
    },
    {
      title: "Terrible!",
    },
  ],
};

describe("<QuickReply />", () => {
  let mockProps;

  function renderElement(props) {
    return render(
      <ThemeProvider>
        <QuickReply {...props} />
      </ThemeProvider>
    );
  }
  window.connect = {
    csmService: {
      addCountMetric: jest.fn().mockImplementation(() => {}),
    }
  }
  beforeEach(() => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = { content: mockQuickReplyContent, addMessage: addMessage };
  });

  it("Style should match the snapshot", () => {
    const mockQuickReply = renderElement(mockProps);
    expect(mockQuickReply).toMatchSnapshot();
  });

  it("Should include all necessary data-testid values for integ/canary tests", () => {
    renderElement(mockProps);

    expect(screen.getByTestId("interactive-quickreply-message-title")).toBeInTheDocument();
    expect(screen.getByTestId("interactive-quickreply-response-section")).toBeInTheDocument();
  });

  it("Should be able to use QuickReply", () => {
    renderElement(mockProps);

    // Renders title
    expect(screen.getByText(mockQuickReplyContent.title)).toBeDefined();

    // Renders each element
    mockQuickReplyContent.elements.forEach(({ title: replyOption }) => {
      expect(screen.getByText(replyOption)).toBeDefined();
    });

    // Can click an option
    const replyToChoose = mockQuickReplyContent.elements[0].title;
    fireEvent.click(screen.getByText(replyToChoose));
    expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
    expect(mockProps.addMessage).toHaveBeenCalledWith({
      text: replyToChoose,
    });
  });

  it("Should render all reply options and be accessible", () => {
    renderElement(mockProps)

    const responseSectionDiv = screen.getByTestId("interactive-quickreply-response-section");
    const buttons = responseSectionDiv.querySelectorAll('button');
    expect(buttons.length).toBeGreaterThan(0);
    buttons.forEach(button => {
      // Check that the button is accessible to screen readers
      expect(button).toHaveAccessibleName();
      button.click();
      expect(button).toBeEnabled();
    });
  });

  it("Should truncate a title over the character limit", () => {
    const { titleCharLimit } = QUICK_REPLY_CONSTRAINTS;

    const longTitle = "LongTitle".repeat(100);
    const truncatedTitle = `${longTitle.substring(0, titleCharLimit)}...`;

    const quickReplyLongTitle = {
      ...mockQuickReplyContent,
      title: longTitle,
    };
    renderElement({ ...mockProps, content: quickReplyLongTitle });
    expect(() => screen.getByText(longTitle)).toThrow(
      "Unable to find an element"
    );
    expect(() => screen.getByText(truncatedTitle)).not.toThrow();
  });

  it("Should truncate a picker option over the character limit", () => {
    const { replyOptionCharLimit } = QUICK_REPLY_CONSTRAINTS;

    const longPickerOption = "LongPickerOption".repeat(100);
    const truncatedPickerOption = `${longPickerOption.substring(0, replyOptionCharLimit)}...`;
    const shortPickerOption = "Under the char limit!";

    const quickReplyLongOption = {
      ...mockQuickReplyContent,
      elements: [{ title: shortPickerOption }, { title: longPickerOption }],
    };

    renderElement({ ...mockProps, content: quickReplyLongOption });
    expect(() => screen.getByText(longPickerOption)).toThrow(
      "Unable to find an element"
    );
    expect(() => screen.getByText(truncatedPickerOption)).not.toThrow();
    expect(() => screen.getByText(shortPickerOption)).not.toThrow();

    // Should still send entire title to backend, instead of returning 'my selectio...'
    fireEvent.click(screen.getByText(truncatedPickerOption));
    expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
    expect(mockProps.addMessage).toHaveBeenCalledWith({
      text: longPickerOption,
    });
  });
});

describe("QuickReply XSS Mitigation", () => {
  let mockProps;
  const mockQuickReplyXSSContent = {
    title: 'Title with script <script>alert("XSS attack!");</script>',
    elements: [
      {
        title: '<a href="javascript:alert(\'XSS attack!\')">Click me</a>',
      },
      {
        title: '<input type="text" value="XSS attack!" onfocus="alert(\'XSS attack!\');">',
      },
      {
        title: '<div data-value="<img src=x onerror=alert(\'XSS attack!\')>"></div>',
      },
      {
        title: '<div style="background-image: url(\'javascript:alert(\'XSS attack!\');\')"></div>',
      }
    ],
  };
  beforeEach(() => {
    mockProps = {
      content: mockQuickReplyXSSContent,
      addMessage: jest.fn(),
      templateType: InteractiveMessageType.QUICK_REPLY,
    };
  });
  function renderElement(props) {
    return render(
      <ThemeProvider>
        <QuickReply {...props} />
      </ThemeProvider>
    );
  }

  it("should use DOMPurify to mitigate malicious XSS input", () => {
    renderElement(mockProps);
    expect(screen.getByText("Title with script")).toBeDefined();
    expect(screen.getByText("<a>Click me</a>", { exact: false })).toBeDefined();
    expect(screen.getByText("<input value=\"XSS attack!\" type=\"text\">", { exact: false })).toBeDefined();
    expect(screen.getByText("<div data-value=\"<img src=x onerror=alert('XSS attack!')>\"></div>", { exact: false })).toBeDefined();
    expect(screen.getByText("<div style=\"background-image: url('javascript:alert('XSS attack!');')\"></div>", { exact: false })).toBeDefined();
  });
});