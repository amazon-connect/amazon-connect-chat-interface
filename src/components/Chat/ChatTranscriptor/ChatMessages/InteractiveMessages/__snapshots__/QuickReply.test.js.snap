// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuickReply /> Style should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c2 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c2:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c2:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c2:active {
  border-color: #fff #fff #fff;
}

.c2:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c2:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c2 > img,
.c2 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c2:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c2:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c2:active {
  border-color: #fff #fff #fff;
}

.c2:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c2:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 {
  border: 1px solid #DDDDDD;
  border-radius: 6px;
  background: #EDEDED;
  padding: 14px;
  position: relative;
}

.c0:after {
  display: block;
  content: " ";
  position: absolute;
  left: -6px;
  bottom: 4px;
  border-radius: 2px;
  border-left: 10px solid transparent;
  border-right: 12px solid transparent;
  border-bottom: 9px solid #EDEDED;
}

.c0 button {
  cursor: pointer;
  border: 1px solid #DDDDDD;
}

.c0 button:hover:enabled {
  color: #fff;
  background: #06adcb;
}

.c1 {
  padding: 14px 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 6px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c3 {
  padding: 4px 14px;
  text-align: left;
  cursor: pointer;
  border: 2px solid #077398;
  border-radius: 2px;
  white-space: pre-wrap;
  font-weight: bold;
  font-size: 0.875rem;
  line-height: 22px;
  font-weight: 600;
  font-style: normal;
  color: #077398;
  background: #FFFFFF;
  text-shadow: none;
  box-shadow: none;
}

.c3:hover,
.c3:focus {
  background: #077398;
  color: #FFFFFF;
  border: 2px solid #077398 !important;
}

<body>
    <div>
      <div
        class="c0"
        data-testid="interactive-quickreply-message-title"
      >
        <p
          style="margin: 0px;"
        >
          How was your experience?
        </p>
      </div>
      <div
        class="c1"
        data-testid="interactive-quickreply-response-section"
      >
        <button
          class="c2 c3"
          data-selected="false"
          type="default"
        >
          Great!
        </button>
        <button
          class="c2 c3"
          data-selected="false"
          type="default"
        >
          Good
        </button>
        <button
          class="c2 c3"
          data-selected="false"
          type="default"
        >
          Ok
        </button>
        <button
          class="c2 c3"
          data-selected="false"
          type="default"
        >
          Poor
        </button>
        <button
          class="c2 c3"
          data-selected="false"
          type="default"
        >
          Terrible!
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-kgAjT jelqub"
      data-testid="interactive-quickreply-message-title"
    >
      <p
        style="margin: 0px;"
      >
        How was your experience?
      </p>
    </div>
    <div
      class="sc-dNLxif ijXeIm"
      data-testid="interactive-quickreply-response-section"
    >
      <button
        class="sc-hMqMXs jEknEr sc-jqCOkK dEJQyU"
        data-selected="false"
        type="default"
      >
        Great!
      </button>
      <button
        class="sc-hMqMXs jEknEr sc-jqCOkK dEJQyU"
        data-selected="false"
        type="default"
      >
        Good
      </button>
      <button
        class="sc-hMqMXs jEknEr sc-jqCOkK dEJQyU"
        data-selected="false"
        type="default"
      >
        Ok
      </button>
      <button
        class="sc-hMqMXs jEknEr sc-jqCOkK dEJQyU"
        data-selected="false"
        type="default"
      >
        Poor
      </button>
      <button
        class="sc-hMqMXs jEknEr sc-jqCOkK dEJQyU"
        data-selected="false"
        type="default"
      >
        Terrible!
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
